packages/nimble-components/src/rich-text-mention/base/models/mention-validator.ts

```diff
@@ -0,0 +1,109 @@
+export interface RichTextMentionValidity extends ValidityObject {}
+
+type FlagNames<T> = T extends readonly (infer U)[] ? U : never;
+
+export const baseValidityFlagNames = [
+    'unsupportedMappingType',
+    'duplicateMappingMentionHref',
+    'missingMentionHrefValue',
+    'unsupportedMentionHrefValue',
+    'missingPatternAttribute'
+] as const;
+
+/**
+ * Validator for RichTextMention
+ */
+export class RichTextMentionValidator<
+    ValidityFlagNames extends readonly string[]
+> extends Validator<typeof baseValidityFlagNames | ValidityFlagNames> {
+    public constructor(
+        private readonly mentionInternals: MentionInternals<unknown>,
+        configValidityKeys: ValidityFlagNames,
+        private readonly supportedMappingElements: readonly (typeof Mapping<unknown>)[]
+    ) {
+        super(configValidityKeys);
+    }
+
+    public validate(
+        mappings: Mapping<unknown>[],
+        pattern: string | undefined
+    ): void {
+        this.untrackAll();
+        const mentionHrefs = mappings.map(mapping => mapping.key);
+        this.validateMappingTypes(mappings);
+        this.validateNoMissingMentionHref(mappings);
+        this.validateUniqueMentionHref(mentionHrefs);
+        this.validatePattern(pattern);
+        this.validateHref(mentionHrefs, pattern);
+    }
+
+    /**
+     * @returns an object containing flags for various ways the configuation can be invalid
+     */
+    public getValidity(): RichTextMentionValidity {
+        return this.getValidationFlags();
+    }
+
+    /**
+     * Sets a particular validity condition flag's value, e.g. "hasInvalidFooValue" = true
+     */
+    protected setConditionValue(
+        name: FlagNames<typeof baseValidityFlagNames | ValidityFlagNames>,
+        isInvalid: boolean
+    ): void {
+        if (isInvalid) {
+            this.track(name);
+        } else {
+            this.untrack(name);
+        }
+        this.updateMentionInternalsFlag();
+    }
+
+    private updateMentionInternalsFlag(): void {
+        this.mentionInternals.validConfiguration = this.isValid();
+    }
+
+    private validatePattern(pattern: string | undefined): void {
+        this.setConditionValue(
+            'missingPatternAttribute',
+            pattern === undefined
+        );
+    }
+
+    private validateMappingTypes(mappings: Mapping<unknown>[]): void {
+        const valid = mappings.every(mapping => this.supportedMappingElements.some(
+            mappingClass => mapping instanceof mappingClass
+        ));
+        this.setConditionValue('unsupportedMappingType', !valid);
+    }
+
+    private validateUniqueMentionHref(
+        mentionHref: (string | undefined | unknown)[]
+    ): void {
+        const invalid = new Set(mentionHref).size !== mentionHref.length;
+        this.setConditionValue('duplicateMappingMentionHref', invalid);
+    }
+
+    private validateNoMissingMentionHref(mappings: Mapping<unknown>[]): void {
```

Comment:
Could you change this to accept `mappingHrefs` like other helpers that only access the mapping keys?

Reply:
Updated